speaker_verification/denoise.py

```diff
@@ -1,130 +1,44 @@
-import math
-import numpy as np
 import os
-import wave
-
-
-def nextpow2(n):
-    return np.ceil(np.log2(np.abs(n))).astype("long")
-
-
-def berouti(SNR):
-    if -5.0 <= SNR <= 20.0:
-        a = 4 - SNR * 3 / 20
-    else:
-        if SNR < -5.0:
-            a = 5
-        if SNR > 20:
-            a = 1
-    return a
-
+import sys
 
-def berouti1(SNR):
-    if -5.0 <= SNR <= 20.0:
-        a = 3 - SNR * 2 / 20
-    else:
-        if SNR < -5.0:
-            a = 4
-        if SNR > 20:
-            a = 1
-    return a
-
-
-def find_index(x_list):
-    index_list = []
-    for i in range(len(x_list)):
-        if x_list[i] < 0:
-            index_list.append(i)
-    return index_list
-
-
-def denoise(audio_file, input_path, output_path):
-    file_name = os.path.basename(audio_file).split(".")[0]
-    input_file_path = os.path.join(input_path, audio_file)
-    output_file_path = os.path.join(output_path, file_name)
-
-    with wave.open(input_file_path, 'r') as f:
-        # (nchannels, sampwidth, framerate, nframes, comptype, compname)
-        params = f.getparams()
-        nchannels, sampwidth, framerate, nframes = params[:4]
-        str_data = f.readframes(nframes)
+import noisereduce as nr
+import numpy as np
+from scipy.io import wavfile
 
-    x = np.frombuffer(str_data, dtype=np.short)
-    len_ = 20 * framerate // 1000
-    len1 = len_ * 50 // 100
-    # The percentage of the frame that the window overlaps
-    len2 = len_ - len1
-    Thres = 3
-    Expnt = 2.0
-    beta = 0.002
-    G = 0.9
-    win = np.hamming(len_)
-    # normalization gain for overlap+add with 50% overlap
-    winGain = len2 / sum(win)
 
-    # Noise magnitude calculations - assuming that the first 5 frames is noise/silence
-    nFFT = 2 * 2 ** (nextpow2(len_))
-    noise_mean = np.zeros(nFFT)
+def denoise(input_file: str) -> (np.ndarray, int):
+    """
+    Denoise audio file using noisereduce.
+    """
 
-    j = 0
-    for k in range(1, 6):
-        noise_mean = noise_mean + abs(np.fft.fft(win * x[j: j + len_], nFFT))
-        j = j + len_
-    noise_mu = noise_mean / 5
+    # TODO: Read other audio types beside .wav
+    # Read audio data
+    rate, data = wavfile.read(input_file)

```

Comment:
In fact the best option maybe to use [`librosa`](https://librosa.org/doc/latest/generated/librosa.load.html#librosa.load) which we've inherited into the repo from the `deep-speaker` dependencies, which seems to utilise the same codec detection as `soundfile`. I think it might be good to use this existing dependency rather than adding a new one.
You could even reuse [`audio_read`](https://github.com/OnTrack-UG-Squad/speaker-verification/blob/development/speaker_verification/deep_speaker/audio.py#L41) from `audio.py`.

Reply:
@aidanjgriffiths It's very hard to rewrite the denoise to use librosa, @MoeHdaib and I agreed that it's far easiest to use noisereduce. Because it will eliminate a lot of unnecessary code we might have to maintain.